client/order/orderbook.go

```diff
@@ -0,0 +1,226 @@
+		}
+
+		for idx := len(ordersAtPrice) - 1; idx >= 0; idx-- {
+			if bytes.Equal(ordersAtPrice[idx].OrderID, order.OrderID) {
+				ordersAtPrice = append(ordersAtPrice[:idx], ordersAtPrice[idx+1:]...)
+				ob.buys[order.Rate] = ordersAtPrice
+				break
+			}
+		}
+
+	case msgjson.SellOrderNum:
+		ordersAtPrice, ok := ob.sells[order.Rate]
+		if !ok {
+			return fmt.Errorf("no sell orders found at price %v", order.Rate)
+		}
+
+		for idx := len(ordersAtPrice) - 1; idx >= 0; idx-- {
+			if bytes.Equal(ordersAtPrice[idx].OrderID, order.OrderID) {
+				ordersAtPrice = append(ordersAtPrice[:idx], ordersAtPrice[idx+1:]...)
+				ob.sells[order.Rate] = ordersAtPrice
+				break
+			}
+		}
+	}
+
+	delete(ob.orders, order.OrderID.String())
+
+	return nil
+}
+
+// Initialize synchronizes a client tracked order book with an order
+// book snapshot.
+func (ob *OrderBook) Initialize(snapshot *msgjson.OrderBook) error {
+	ob.mtx.Lock()
+	defer ob.mtx.Unlock()
+
+	if ob.synced {
+		return fmt.Errorf("reinitializing a synced order book")
+	}
+
+	ob.orders = make(map[string]*msgjson.BookOrderNote, len(snapshot.Orders))
+	ob.buys = make(map[uint64][]*msgjson.BookOrderNote)
+	ob.sells = make(map[uint64][]*msgjson.BookOrderNote)
+
+	for _, order := range snapshot.Orders {
+		ob.addOrder(order)
+	}
+
+	ob.seq = snapshot.Seq
+	ob.marketID = snapshot.MarketID
+	ob.synced = true
+
+	return nil
+}
+
+// Book adds a new order to the order book.
+func (ob *OrderBook) Book(order *msgjson.BookOrderNote) error {
+	ob.mtx.Lock()
+	defer ob.mtx.Unlock()
+
+	if !ob.synced {
+		return fmt.Errorf("order book is unsynced")
+	}
+
+	if ob.marketID != order.MarketID {
+		return fmt.Errorf("market id mismatch, expected %s got %s",
+			ob.marketID, order.MarketID)
+	}
+
+	expectedSeq := ob.seq + 1
+	if expectedSeq != order.Seq {
+		ob.synced = false
+		return fmt.Errorf("order book out of sync, expected %d got %d",
+			expectedSeq, order.Seq)
+	}
+
+	ob.seq++
+	ob.addOrder(order)
+	return nil
+}
+
+// Unbook removes an order from the order book.
+func (ob *OrderBook) Unbook(unbookOrder *msgjson.UnbookOrderNote) error {
+	ob.mtx.Lock()
+	defer ob.mtx.Unlock()
+
+	if !ob.synced {
+		return fmt.Errorf("order book is unsynced")
+	}
+
+	if ob.marketID != unbookOrder.MarketID {
+		return fmt.Errorf("market id mismatch, expected %s got %s",
+			ob.marketID, unbookOrder.MarketID)
+	}
+
+	order, ok := ob.orders[unbookOrder.OrderID.String()]
+	if !ok {
+		return fmt.Errorf("no order found with id %v",
+			unbookOrder.OrderID.String())
+	}
+
+	expectedSeq := ob.seq + 1
+	if expectedSeq != order.Seq {
+		ob.synced = false
+		return fmt.Errorf("order book out of sync, expected %v got %v",
+			expectedSeq, order.Seq)
+	}
+
+	ob.seq++
+	err := ob.removeOrder(order)
+	if err != nil {
+		ob.synced = false
+		return err
+	}
+	return nil
+}
+
+// Synced returns the sync state of the order book.
+func (ob *OrderBook) Synced() bool {
+	ob.mtx.RLock()
+	defer ob.mtx.RUnlock()
+	return ob.synced
+}
```

Comment:
I usually don't like to write these types of state synchronization methods until I know they're needed. The `OrderBook` consumer should probably try resyncing on any error, so an additional call to `Synced` is like superfluous, since all exported methods return an error when `ob.synced` is `false` anyway.